src/main/java/org/opensearch/neuralsearch/query/SparseEncodingQueryBuilder.java

```diff
@@ -0,0 +1,306 @@
+    @VisibleForTesting
+    static final ParseField QUERY_TEXT_FIELD = new ParseField("query_text");
+    @VisibleForTesting
+    static final ParseField MODEL_ID_FIELD = new ParseField("model_id");
+
+    private static MLCommonsClientAccessor ML_CLIENT;
+
+    public static void initialize(MLCommonsClientAccessor mlClient) {
+        SparseEncodingQueryBuilder.ML_CLIENT = mlClient;
+    }
+
+    private String fieldName;
+    private Map<String, Float> queryTokens;
+    private String queryText;
+    private String modelId;
+    private Supplier<Map<String, Float>> queryTokensSupplier;
+
+    public SparseEncodingQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        // we don't have readOptionalMap or write, need to do it manually
+        if (in.readBoolean()) {
+            this.queryTokens = in.readMap(StreamInput::readString, StreamInput::readFloat);
+        }
+        this.queryText = in.readOptionalString();
+        this.modelId = in.readOptionalString();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        if (null != queryTokens) {
+            out.writeBoolean(true);
+            out.writeMap(queryTokens, StreamOutput::writeString, StreamOutput::writeFloat);
+        } else {
+            out.writeBoolean(false);
+        }
+        out.writeOptionalString(queryText);
+        out.writeOptionalString(modelId);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder xContentBuilder, Params params) throws IOException {
+        xContentBuilder.startObject(NAME);
+        xContentBuilder.startObject(fieldName);
+        if (null != queryTokens) xContentBuilder.field(QUERY_TOKENS_FIELD.getPreferredName(), queryTokens);
+        if (null != queryText) xContentBuilder.field(QUERY_TEXT_FIELD.getPreferredName(), queryText);
+        if (null != modelId) xContentBuilder.field(MODEL_ID_FIELD.getPreferredName(), modelId);
+        printBoostAndQueryName(xContentBuilder);
+        xContentBuilder.endObject();
+        xContentBuilder.endObject();
+    }
+
+    /**
+     * The expected parsing form looks like:
+     * {
+     *  "SAMPLE_FIELD": {
+     *    "query_tokens": {
+     *        "token_a": float,
+     *        "token_b": float,
+     *        ...
+     *    }
+     *  }
+     * }
+     * or
+     *  "SAMPLE_FIELD": {
+     *    "query_text": "string",
+     *    "model_id": "string"
+     *  }
+     *
+     */
+    public static SparseEncodingQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        SparseEncodingQueryBuilder sparseEncodingQueryBuilder = new SparseEncodingQueryBuilder();
+        if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(), "First token of " + NAME + "query must be START_OBJECT");
```

Comment:
Do we need this parser.getTokenLocation() since we already explicitly identifying this is the First token?

Reply:
This is the first token of sparse_encoding query object, but its location is uncertain. The query object may be wrapped by other query objects.